heaps/heap_sort.py

```diff
@@ -1,8 +1,77 @@
+def swap(list, index_1, index_2):
+    """ Swaps two elements in a list"""
+    list[index_1], list[index_2] = list[index_2], list[index_1]
+
+
+def find_left_child(list, index):
+    return 2 * index + 1
+
+
+def find_right_child(list, index):
+    return 2 * index + 2
+
+
+def heap_down(list, index, last_node_index):
+    """ Takes an index and moves it down the heap if it's smaller
+        than it's child node.
+        last_node_index - defines the bound b/w sorted and unsorted
+        part of array.
+    """
+
+    length = last_node_index
+
+    while True:
+
+        left = find_left_child(list, index)
+        right = find_right_child(list, index)
+
+        if right < length:
+            # both left and right children exist
+            # swap with the biggest of them
+            if list[index] >= max(list[left], list[right]):
+                break
+            elif list[left] < list[right]:
+                swap(list, index, right)
+                index = right
+            else:
+                swap(list, index, left)
+                index = left
+        elif left < length:
+            # no right child
+            # check if the only child is bigger:
+            # yes -> swap
+            # no -> break, because we reached the end of the heap
+            if list[left] > list[index]:
+                swap(list, index, left)
+                index = left
+            else:
+                break
+        else:
+            break
 
 
 def heap_sort(list):
```

Comment:
👍  Really nice work building an in-place heapsort, using the given list as the internal array for the heap.  Well done.